### PR TITLE
examples: Add e2e testing with dAppwright

### LINKED ISSUE #417: Improve CI to support UI MetaMask testing
Check out, what's currently hot in the crypto space to for testing your dApp behavior with brave/metamask. Find any relevant sapphire-paratime issues related to this (https://github.com/oasisprotocol/sapphire-paratime/issues/242 ?).

cc the Wallet/dApp team @lubej @lukaw3d

### PR DESCRIPTION
**Pull request by aefhm**
## Description

Actually close #417 this time.

## TODO

- [x] check screen shots on [failure](https://github.com/oasisprotocol/sapphire-paratime/actions/runs/11658313279/job/32457179468?pr=451) is [downloadable](https://github.com/oasisprotocol/sapphire-paratime/actions/runs/11658313279/artifacts/2139247896)
- [x] improve testing documentation
- [x] rethink testing runtime cost // technically not the longest check yet

### CODE REVIEW

File: .github/workflows/ci-playwright.yaml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        working-directory: examples/wagmi-v2
+        run: xvfb-run pnpm test
+
+      - name: 'Upload playwright test-results'
+        if: ${{ failure() }}
+        uses: actions/upload-artifact@v4
+        with:
+          name: playwright-test-results
+          path: examples/wagmi-v2/test-results
+          retention-days: 5
```
Comment (aefhm) [inline .github/workflows/ci-playwright.yaml:84]:
```suggestion
          retention-days: 5

```

File: .github/workflows/ci-playwright.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+          wait-for: 31sec
+          log-output: true
+          log-output-if: true
+          working-directory: clients/js
+      - name: Install dependencies
+        run: pnpm install && pnpm test:setup
+        working-directory: examples/wagmi-v2
+      # - name: Set up playwright tests
+      #   run: npx playwright install --with-deps
+      #   working-directory: examples/wagmi-v2
```
Comment (aefhm) [inline .github/workflows/ci-playwright.yaml:72]:
```suggestion
```

File: examples/wagmi-v2/test/e2e.spec.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  await expect(page.getByText("Contract:")).toBeVisible();
+
+  await page.getByText("Write").click();
+  await wallet.confirmTransaction();
+  await expect(page.getByText("Contract:")).toBeVisible();
+  await expect(page.getByTestId('is-write-enveloped')).toHaveText('encrypted');
+
+  await page.getByText("Read").click();
+  await expect(page.getByTestId("read-result")).not.toBeEmpty();
+});
```
Comment (aefhm) [inline examples/wagmi-v2/test/e2e.spec.ts:58]:
```suggestion
});

```

File: examples/wagmi-v2/playwright.config.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      testDir: './test/',
+      use: {
+        ...devices['Desktop Chrome'],
+      },
+    },
+  ],
+  /* Run your local dev server before starting the tests */
+  webServer: {
+    command: 'pnpm run start:server',
+    url: 'http://localhost:3000',
```
Comment (aefhm) [inline examples/wagmi-v2/playwright.config.ts:41]:
```suggestion
    url: process.env.FRONTEND_URL || "http://localhost:3000/",
```

File: examples/wagmi-v2/test/e2e.spec.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import dappwright, { Dappwright, MetaMaskWallet } from "@tenkeylabs/dappwright";
+
+export const test = baseTest.extend<{
+  context: BrowserContext;
+  wallet: Dappwright;
+}>({
+  context: async ({}, use) => {
+    // Launch context with extension
+    const [wallet, _, context] = await dappwright.bootstrap("", {
+      wallet: "metamask",
```
Comment (aefhm) [inline examples/wagmi-v2/test/e2e.spec.ts:11]: No [Brave](https://github.com/TenKeyLabs/dappwright/blob/d7af3fa0e6aadf28928e9a8b25475cbc3c76cec6/src/wallets/wallets.ts#L2-L3) yet. We'd have to write that a little by hand.

File: examples/wagmi-v2/test/e2e.spec.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  },
+
+  wallet: async ({ context }, use) => {
+    const metamask = await dappwright.getWallet("metamask", context);
+
+    await use(metamask);
+  },
+});
+
+test.beforeEach(async ({ wallet, page }) => {
```
Comment (aefhm) [inline examples/wagmi-v2/test/e2e.spec.ts:35]:
```suggestion
test.beforeEach(async ({ wallet, page }) => {
    // Use first account from seed. dAppwright adds two accounts by default.
```

File: examples/wagmi-v2/test/e2e.spec.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  await wallet.switchAccount(1); // Selector queries as a string
+  await page.bringToFront();
+});
+
+test("deploy contract and send encrypted transaction", async ({ wallet, page }) => {
+  await page.goto("http://localhost:3000");
+
+  await page.getByText("Injected (Sapphire)").click();
+  await wallet.approve();
+  await expect(page.getByText("0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266")).toBeVisible();
```
Comment (aefhm) [inline examples/wagmi-v2/test/e2e.spec.ts:45]: Maybe note that this is the first account address.

File: examples/wagmi-v2/test/e2e.spec.ts (hunk 5 above)
Comment (CedarMist) [reply examples/wagmi-v2/test/e2e.spec.ts:11]: That's ok :D

For now this is awesome!

**Review by CedarMist — APPROVED** [review]

**Pull request merged**